In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
#reading csv file
df_car = pd.read_csv("Car_dheko_data.csv")

In [3]:
#copying the DataFarme
df_copy = df_car.copy()

In [4]:
df_copy.head(2)

Source.Name web-scraper-order  \
0  cardekho_extract(0-2000).csv   1611917819-1662   
1  cardekho_extract(0-2000).csv   1611918361-1902   

                               web-scraper-start-url               full_name  \
0  https://www.cardekho.com/used-car-details/used...         Maruti Alto Std   
1  https://www.cardekho.com/used-car-details/used...  Hyundai Grand i10 Asta   

  selling_price                                     new-price    year  \
0     1.2 Lakh*                                           NaN  2012.0   
1     5.5 Lakh*  New Car (On-Road Price) : Rs.7.11-7.48 Lakh*  2016.0   

  seller_type     km_driven   owner_type fuel_type transmission_type  \
0  Individual  1,20,000 kms  First Owner    Petrol            Manual   
1  Individual    20,000 kms  First Owner    Petrol            Manual   

            mileage         engine          max_power   seats  
0  Mileage19.7 kmpl   Engine796 CC  Max Power46.3 bhp  Seats5  
1  Mileage18.9 kmpl  Engine1197 CC    Max Power82 bhp  Seats5

In [5]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20026 entries, 0 to 20025
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Source.Name            20026 non-null  object 
 1   web-scraper-order      20026 non-null  object 
 2   web-scraper-start-url  20026 non-null  object 
 3   full_name              19980 non-null  object 
 4   selling_price          19980 non-null  object 
 5   new-price              9566 non-null   object 
 6   year                   19980 non-null  float64
 7   seller_type            19980 non-null  object 
 8   km_driven              19980 non-null  object 
 9   owner_type             19980 non-null  object 
 10  fuel_type              19980 non-null  object 
 11  transmission_type      19980 non-null  object 
 12  mileage                19980 non-null  object 
 13  engine                 19921 non-null  object 
 14  max_power              19921 non-null  object 
 15  se

In [6]:
df_copy.describe(include='all')

Source.Name web-scraper-order  \
count                              20026             20026   
unique                                 6             20026   
top     cardekho_extract(5001-10000).csv   1612099971-1963   
freq                                5000                 1   
mean                                 NaN               NaN   
std                                  NaN               NaN   
min                                  NaN               NaN   
25%                                  NaN               NaN   
50%                                  NaN               NaN   
75%                                  NaN               NaN   
max                                  NaN               NaN   

                                    web-scraper-start-url  \
count                                               20026   
unique                                              20026   
top     https://www.cardekho.com/used-car-details/used...   
freq                                                    1   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                     full_name selling_price  \
count                    19980         19980   
unique                    3321          1215   
top     Maruti Swift Dzire VDI     4.5 Lakh*   
freq                       210           467   
mean                       NaN           NaN   
std                        NaN           NaN   
min                        NaN           NaN   
25%                        NaN           NaN   
50%                        NaN           NaN   
75%                        NaN           NaN   
max                        NaN           NaN   

                                             new-price          year  \
count                                             9566  19980.000000   
unique                                            1744           NaN   
top     New Car (On-Road Price) : Rs.12.68-16.80 Lakh*           NaN   
freq                                               144           NaN   
mean                                               NaN   2014.525125   
std                                                NaN      3.249185   
min                                                NaN   1991.000000   
25%                                                NaN   2013.000000   
50%                                                NaN   2015.000000   
75%                                                NaN   2017.000000   
max                                                NaN   2021.000000   

       seller_type   km_driven   owner_type fuel_type transmission_type  \
count        19980       19980        19980     19980             19980   
unique           3        4515            3         5                 2   
top         Dealer  50,000 kms  First Owner    Diesel            Manual   
freq         11973         751        19974      9817             16029   
mean           NaN         NaN          NaN       NaN               NaN   
std            NaN         NaN          NaN       NaN               NaN   
min            NaN         NaN          NaN       NaN               NaN   
25%            NaN         NaN          NaN       NaN               NaN   
50%            NaN         NaN          NaN       NaN               NaN   
75%            NaN         NaN          NaN       NaN               NaN   
max            NaN         NaN          NaN       NaN               NaN   

                 mileage         engine        max_power   seats  
count              19980          19921            19921   19853  
unique               564            173              498      10  
top    

# Removing unwanted fetures

In [7]:
df_copy.nunique()

Source.Name                  6
web-scraper-order        20026
web-scraper-start-url    20026
full_name                 3321
selling_price             1215
new-price                 1744
year                        28
seller_type                  3
km_driven                 4515
owner_type                   3
fuel_type                    5
transmission_type            2
mileage                    564
engine                     173
max_power                  498
seats                       10
dtype: int64

##### From above analysis on number of unique values, we can say that
- web-scraper-order,web-scraper-start-url column has 20026 unique values which is equal to that of no.of rows in our dataset. So we can drop this column, since it doesnot add any information to our dataset.

- The columns which has only less unique values can be considered as categorical columns i.e. columns 'seller_type','owner_type','fule_type','transmission_type' and 'seat'.

- Columns 'full_name', 'selling_price', 'new-price', 'km_driven', 'mileage', 'engine' and 'max_power' are of type object.
- Column 'Year' is numerical column.

#### Dropping "web-scraper-order,web-scraper-start-url and Source.Name" column since we have all unique values

In [8]:
df_copy.drop(['Source.Name','web-scraper-start-url','web-scraper-order'], inplace=True, axis=1)

In [9]:
df_copy.columns

Index(['full_name', 'selling_price', 'new-price', 'year', 'seller_type',
       'km_driven', 'owner_type', 'fuel_type', 'transmission_type', 'mileage',
       'engine', 'max_power', 'seats'],
      dtype='object')

# Converting objects in integer

In [10]:
#defined Empty string
empty_str = ""

#Function to convert string objects containing numerical details

#Method will only extract numerical value as strings and convert it into float
def Object_to_interger(x):
        value = empty_str.join(element for element in str(x) if (element.isdigit() or element =='.'))
        value = pd.to_numeric(value,errors = 'coerce')
        return value      

In [11]:
df_copy['mileage'] = df_copy['mileage'].apply(Object_to_interger)

In [12]:
df_copy['engine'] = df_copy['engine'].apply(Object_to_interger)

In [13]:
df_copy['seats'] = df_copy['seats'].apply(Object_to_interger)

In [14]:
df_copy['max_power'] = df_copy['max_power'].apply(Object_to_interger)

In [15]:
df_copy['km_driven'] = df_copy['km_driven'].apply(Object_to_interger)

In [16]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20026 entries, 0 to 20025
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   full_name          19980 non-null  object 
 1   selling_price      19980 non-null  object 
 2   new-price          9566 non-null   object 
 3   year               19980 non-null  float64
 4   seller_type        19980 non-null  object 
 5   km_driven          19980 non-null  float64
 6   owner_type         19980 non-null  object 
 7   fuel_type          19980 non-null  object 
 8   transmission_type  19980 non-null  object 
 9   mileage            19980 non-null  float64
 10  engine             19890 non-null  float64
 11  max_power          19684 non-null  float64
 12  seats              19852 non-null  float64
dtypes: float64(6), object(7)
memory usage: 2.0+ MB


In [17]:
#Spliting Maker and Model using full_name
tem = df_copy["full_name"].str.upper()
df_copy['full_name'] = tem
new = df_copy["full_name"].str.split(" ", n = 3, expand = True)
  
# making separate first name column from new data frame
df_copy["Maker"]= new[0]
df_copy['Model']= new[1] + " " + new[2]
df_copy.drop('full_name', inplace=True, axis=1)

In [18]:
df_copy.selling_price = df_copy.selling_price.str.replace('[*,,]', '')
df_copy[['selling_price','unit']] = df_copy.selling_price.str.split(expand=True)
df_copy['selling_price'] = df_copy['selling_price'].astype('float64', errors = 'raise')

df_copy.loc[df_copy.unit == "Lakh", 'selling_price'] = df_copy['selling_price']*100000.0
df_copy.loc[df_copy.unit == "Cr", 'selling_price'] = df_copy['selling_price']*10000000.0


df_copy=df_copy.drop(['unit'],axis=1)

df_copy.head()

selling_price                                       new-price    year  \
0       120000.0                                             NaN  2012.0   
1       550000.0    New Car (On-Road Price) : Rs.7.11-7.48 Lakh*  2016.0   
2       215000.0                                             NaN  2010.0   
3       226000.0                                             NaN  2012.0   
4       570000.0  New Car (On-Road Price) : Rs.10.14-13.79 Lakh*  2015.0   

  seller_type  km_driven   owner_type fuel_type transmission_type  mileage  \
0  Individual   120000.0  First Owner    Petrol            Manual    19.70   
1  Individual    20000.0  First Owner    Petrol            Manual    18.90   
2  Individual    60000.0  First Owner    Petrol            Manual    17.00   
3  Individual    37000.0  First Owner    Petrol            Manual    20.92   
4      Dealer    30000.0  First Owner    Diesel            Manual    22.77   

   engine  max_power  seats    Maker               Model  
0   796.0      46.30    5.0   MARUTI            ALTO STD  
1  1197.0      82.00    5.0  HYUNDAI           GRAND I10  
2  1197.0      80.00    5.0  HYUNDAI            I20 ASTA  
3   998.0      67.10    5.0   MARUTI            ALTO K10  
4  1498.0      98.59    5.0     FORD  ECOSPORT 2015-2021

In [20]:
df_copy['new_price1']=df_copy['new-price'].str.lstrip('New Car (On-Road Price) : Rs.')
df_copy.new_price1 = df_copy.new_price1.str.replace('[*,,]', '')

df_copy[['new_price1','unit']] = df_copy.new_price1.str.split(" ",expand=True)

df_copy[['Min new price','Max new price']] = df_copy.new_price1.str.split("-",expand=True)
df_copy['Min new price'] = df_copy['Min new price'].str.replace('[A-Za-z]', '')
df_copy['Max new price'] = df_copy['Max new price'].str.replace('[A-Za-z]', '')

df_copy.drop(['new_price1'],axis=1)
df_copy.head(1)

# Changing datatype into float

df_copy['Max new price'] = df_copy['Max new price'].astype('float64', errors = 'raise')
df_copy['Min new price'] = df_copy['Min new price'].astype('float64', errors = 'raise')

# Converting cost price to appropriate units

df_copy.loc[df_copy.unit == "Lakh", 'Min new price'] = df_copy['Min new price']*100000.0
df_copy.loc[df_copy.unit == "Lakh", 'Max new price'] = df_copy['Max new price']*100000.0

df_copy.loc[df_copy.unit == "Cr", 'Min new price'] = df_copy['Min new price']*100000.0
df_copy.loc[df_copy.unit == "Cr", 'Max new price'] = df_copy['Max new price']*10000000.0

df_copy.drop(['unit','new_price1'],axis=1, inplace=True)

In [21]:
#creating age column

# importing the date class datetime module
from datetime import date
# creating the date object of today's date
current_date = date.today()    
    
df_copy['Age'] = df_copy['year'].apply(lambda x : current_date.year - x)

df_copy.drop('year', inplace=True, axis=1)

In [22]:
df_copy.head()

selling_price                                       new-price seller_type  \
0       120000.0                                             NaN  Individual   
1       550000.0    New Car (On-Road Price) : Rs.7.11-7.48 Lakh*  Individual   
2       215000.0                                             NaN  Individual   
3       226000.0                                             NaN  Individual   
4       570000.0  New Car (On-Road Price) : Rs.10.14-13.79 Lakh*      Dealer   

   km_driven   owner_type fuel_type transmission_type  mileage  engine  \
0   120000.0  First Owner    Petrol            Manual    19.70   796.0   
1    20000.0  First Owner    Petrol            Manual    18.90  1197.0   
2    60000.0  First Owner    Petrol            Manual    17.00  1197.0   
3    37000.0  First Owner    Petrol            Manual    20.92   998.0   
4    30000.0  First Owner    Diesel            Manual    22.77  1498.0   

   max_power  seats    Maker               Model  Min new price  \
0      46.30    5.0   MARUTI            ALTO STD            NaN   
1      82.00    5.0  HYUNDAI           GRAND I10       711000.0   
2      80.00    5.0  HYUNDAI            I20 ASTA            NaN   
3      67.10    5.0   MARUTI            ALTO K10            NaN   
4      98.59    5.0     FORD  ECOSPORT 2015-2021      1014000.0   

   Max new price   Age  
0            NaN   9.0  
1       748000.0   5.0  
2            NaN  11.0  
3            NaN   9.0  
4      1379000.0   6.0

In [23]:
df_copy.nunique()

selling_price        1215
new-price            1744
seller_type             3
km_driven            4499
owner_type              3
fuel_type               5
transmission_type       2
mileage               533
engine                171
max_power             490
seats                   9
Maker                  41
Model                 993
Min new price        1136
Max new price        1160
Age                    28
dtype: int64

In [24]:
df_copy.describe().apply(lambda s: s.apply(lambda x: format(x, 'f')))

selling_price       km_driven       mileage        engine  \
count     19980.000000    19980.000000  19980.000000  19890.000000   
mean     739206.602302    58244.878979     19.347219   1476.448441   
std      910308.808934    51725.090445      4.620053    521.305479   
min       25000.000000      100.000000      0.000000      0.000000   
25%      340000.000000    31164.250000     16.800000   1197.000000   
50%      520000.000000    52000.000000     19.160000   1248.000000   
75%      785000.000000    74000.000000     22.320000   1582.000000   
max    39500000.000000  3800000.000000    120.000000   6752.000000   

          max_power         seats   Min new price    Max new price  \
count  19684.000000  19852.000000     9566.000000      9280.000000   
mean      99.250751      5.319917  1428429.855739   2021487.823276   
std       43.945637      0.837668  1614605.796581   3153493.408475   
min        5.000000      2.000000   101000.000000    436000.000000   
25%       73.940000      5.000000   653000.000000    884750.000000   
50%       86.800000      5.000000   900500.000000   1284000.000000   
75%      113.420000      5.000000  1310000.000000   1811000.000000   
max      626.000000     14.000000  9880000.000000  91100000.000000   

                Age  
count  19980.000000  
mean       6.474875  
std        3.249185  
min        0.000000  
25%        4.000000  
50%        6.000000  
75%        8.000000  
max       30.000000

In [25]:
df_car.describe().apply(lambda s: s.apply(lambda x: format(x, 'f')))

year
count  19980.000000
mean    2014.525125
std        3.249185
min     1991.000000
25%     2013.000000
50%     2015.000000
75%     2017.000000
max     2021.000000

In [26]:
df_copy.dtypes

selling_price        float64
new-price             object
seller_type           object
km_driven            float64
owner_type            object
fuel_type             object
transmission_type     object
mileage              float64
engine               float64
max_power            float64
seats                float64
Maker                 object
Model                 object
Min new price        float64
Max new price        float64
Age                  float64
dtype: object

In [27]:
round(df_copy.isnull().sum()/len(df_copy)*100 ,2)

selling_price         0.23
new-price            52.23
seller_type           0.23
km_driven             0.23
owner_type            0.23
fuel_type             0.23
transmission_type     0.23
mileage               0.23
engine                0.68
max_power             1.71
seats                 0.87
Maker                 0.23
Model                 0.24
Min new price        52.23
Max new price        53.66
Age                   0.23
dtype: float64

In [28]:
df_copy.dropna(subset = ['selling_price','seller_type','km_driven','owner_type','fuel_type','transmission_type','mileage','engine','max_power','seats','Maker','Model','Age'],inplace=True)

In [29]:
round(df_copy.isnull().sum()/len(df_copy)*100 ,2)

selling_price         0.00
new-price            51.75
seller_type           0.00
km_driven             0.00
owner_type            0.00
fuel_type             0.00
transmission_type     0.00
mileage               0.00
engine                0.00
max_power             0.00
seats                 0.00
Maker                 0.00
Model                 0.00
Min new price        51.75
Max new price        53.19
Age                   0.00
dtype: float64

In [30]:
col_order=['Maker','Model','Min new price','Max new price','Age','km_driven','seller_type','owner_type','fuel_type','transmission_type','mileage','engine','max_power','seats','selling_price']
df_copy=df_copy[col_order]
df_copy.head(1)

Maker     Model  Min new price  Max new price  Age  km_driven seller_type  \
0  MARUTI  ALTO STD            NaN            NaN  9.0   120000.0  Individual   

    owner_type fuel_type transmission_type  mileage  engine  max_power  seats  \
0  First Owner    Petrol            Manual     19.7   796.0       46.3    5.0   

   selling_price  
0       120000.0

In [31]:
# Exporting it to a new file
df_copy.to_csv('processed_data.csv',index = None,header=True)